DynamicStoragePiles/RestrictContainers.cs

```diff
@@ -0,0 +1,298 @@
+﻿using System.Collections.Generic;
+using HarmonyLib;
+
+namespace DynamicStoragePiles {
+
+    /// <summary>
+    ///     Class to patch Valheim inventory methods so that the specified
+    ///     containers can only contain items of a single type.
+    /// </summary>
+    [HarmonyPatch(typeof(Inventory))]
+    internal static class RestrictContainers {
+        private static string _loadingContainer;
+        private static string _targetContainer;
+        private static string _allowedItem;
+        private static Dictionary<string, string> _allowedItemsByContainer = new Dictionary<string, string>();
+
+        /// <summary>
+        ///     Set which containers to restrict and which item is allowed to be placed in each one.
+        /// </summary>
+        /// <param name="allowedItemByContainer"></param>
+        public static void SetContainerRestrictions(Dictionary<string, string> allowedItemByContainer) {
+            _allowedItemsByContainer = allowedItemByContainer;
+        }
+
+        /// <summary>
+        ///     Checks if adding item to inventory should be blocked based on
+        ///     if the container and item names are restricted. Also notifies
+        ///     the player of why adding the item is not allowed.
+        /// </summary>
+        /// <param name="inventory"></param>
+        /// <param name="item"></param>
+        /// <returns></returns>
+        private static bool CanAddItem(Inventory inventory, ItemDrop.ItemData item) {
+            if (inventory == null || item == null) { return false; }
+
+            Jotunn.Logger.LogDebug("CanAddItem");
+            Jotunn.Logger.LogDebug($"Add to: {inventory.m_name}");
+            Jotunn.Logger.LogDebug($"Item: {item.PrefabName()}");
+
+            // Return early if this check is being called while loading a container.
+            // Don't want to delete "non-allowed" items that were put in the container
+            // while the restrictions were disabled.
+            if (!string.IsNullOrEmpty(_loadingContainer) && inventory.m_name == _loadingContainer) {
```

Comment:
```suggestion
            if (inventory.m_name == _loadingContainer) {
```

the string.IsNullOrEmpty check isn't needed, since two strings are compared